lib/core/Concurrency.cc

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <core/Concurrency.h>
+
+#include <core/CLogger.h>
+
+#include <boost/make_unique.hpp>
+
+#include <memory>
+#include <thread>
+
+namespace ml {
+namespace core {
+namespace {
+class CExecutorHolder {
+public:
+    CExecutorHolder()
+        : m_ThreadPoolSize{0}, m_Executor{boost::make_unique<transwarp::sequential>()} {}
+
+    static CExecutorHolder makeThreadPool(std::size_t threadPoolSize) {
+        if (threadPoolSize == 0) {
+            threadPoolSize = std::thread::hardware_concurrency();
+        }
+
+        if (threadPoolSize > 0) {
+            try {
+                return CExecutorHolder{threadPoolSize};
+            } catch (const std::exception& e) {
+                LOG_ERROR(<< "Failed to create thread pool: " << e.what());
+            }
+        } else {
+            LOG_WARN(<< "Failed to determine hardware concurrency");
+        }
+
+        // We'll just have to execute in the main thread.
+        return CExecutorHolder{};
+    }
+
+    executor& executor() { return *m_Executor; }
+    std::size_t threadPoolSize() const { return m_ThreadPoolSize; }
+
+private:
+    CExecutorHolder(std::size_t threadPoolSize)
+        : m_ThreadPoolSize{threadPoolSize},
+          m_Executor{boost::make_unique<transwarp::parallel>(threadPoolSize)} {}
+
+private:
+    std::size_t m_ThreadPoolSize;
+    std::unique_ptr<core::executor> m_Executor;
+};
+
+CExecutorHolder singletonExecutor;
+}
+
+void startDefaultAsyncExecutor(std::size_t threadPoolSize) {
+    singletonExecutor = CExecutorHolder::makeThreadPool(threadPoolSize);
```

Comment:
looks unsafe to me, what if this is called more than once and worse, while it the old executor is in use.

Reply:
The assignment will destroy the old executor, which will wait for it to finish processing.

I agree the method is not thread safe though (nor is the corresponding `stop`).  If the work of these two methods were locked then they would be.

I think there's an assumption these methods will only be used in `main()` and unit tests.  It's probably worth documenting this.

Reply:
It is commented as such in the header. I thought about maybe making this call_once, but it makes writing test code hard.

One definitely shouldn't be calling this throughout a running data frame analysis and I think locking this makes the intention ambiguous.

Maybe production code could use call_once and we have "test only" functions to start and stop.

Reply:
I'm happy to leave these as not thread safe and just have comments to say only use in `main()` and unit tests.

Even though it's documented in the header it might be worth adding a comment in the `.cc` file to say something like "not thread safe - see header for reasons" just to avoid any danger of someone deciding to make them thread safe in the future.

Reply:
what I meant with unsafe: 
```
startDefaultAsyncExecutor(10);
e_ref = executor(); // gets a reference to the executor
e_ref.doSomething() // ok, this is safe
startDefaultAsyncExecutor(10); // invalidates e, but waits for the e to finish
e_ref.doSomething() // the ref is invalid, therefore this is doomed!
```

I thought of an atomic bool as guard in ExecutorHolder, on the other side, this doesn't solve the problem completely. The only way to make it bullet proof would be to proxy the executor and get rid of `executor()` - which makes it super complex. "test only" sounds good to me.